### PR TITLE
dev/core#735 Do not include product in search results if site has none

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Suppress products from contribution queries where no products exist on the site.

This is mostly for performance but it does get rid of the 'Premium' column for sites that don't use it. Many people WOULD like to replace it but they have varied alternatives so hooks / future funded leap into configurability are the best place fotr that.

Before
----------------------------------------
screenshot 2019-02-19 17 27 30

After
----------------------------------------
No change if someone has purchase a product off the site & there is a db record of it - otherwise

screenshot 2019-02-19 17 28 17

Technical Details
----------------------------------------
@colemanw this turned out to be surprisingly easy due to earlier cleanup by @monishdeb  & myself. Am expecting some test fails but I'll update the tests at that point

Comments
----------------------------------------
https://lab.civicrm.org/dev/core/issues/735

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@colemanw wanna check this?  passing now

File: CRM/Contribute/BAO/Query.php [hunk 1]
```diff
@@ -773,23 +773,33 @@ public static function selectorReturnProperties($queryParams) {
...
+   * Do any products exist in this site's database.
+   *
+   * @return bool
+   */
+  public static function isSiteHasProducts() {
+    if (!isset(\Civi::$statics[__CLASS__]['has_products'])) {
+      \Civi::$statics[__CLASS__]['has_products'] = (bool) CRM_Core_DAO::singleValueQuery('SELECT id FROM civicrm_contribution_product LIMIT 1');
+    }
+    return \Civi::$statics[__CLASS__]['has_products'];
+  }
```
Comment (colemanw) [inline CRM/Contribute/BAO/Query.php]: The doc for this function should contain this:

hasproduct

Comment (eileenmcnaughton) [reply CRM/Contribute/BAO/Query.php]: lol

**Review by colemanw — CHANGES_REQUESTED** [review]
I tested this out on the demo site by going to **Search -> Find Contributions** and found a couple issues:

- After clicking "Search" (with no criteria) I found no Premium column in the results even though a premium exists in the system (Coffee Mug).
- I tried the search again selecting 1 search criteria: Premium = Coffee Mug. This caused an error:
> Notice: Undefined index: contribution_product_id in CRM_Contribute_BAO_Query::whereClauseSingle() (line 464 of /home/jenkins/bknix-dfl/build/core-13638-6jcaq/sites/all/modules/civicrm/CRM/Contribute/BAO/Query.php).

**Comment by eileenmcnaughton** [comment]
@colemanw I agree the enotice needs fixing but note that I used ```SELECT id FROM civicrm_contribution_product LIMIT 1``` to determine if products exist - mostly because it seems lots of sites have created a product at some point & then abandoned the concept so having actually sold one seemed to show intent to me

**Comment by eileenmcnaughton** [comment]
so I updated it so not show the premium field if the kitty is sad - I can update the criteria to a different table if you think it would be better - I figured it will never return results unless contribution_product has rows regardless

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by colemanw** [comment]
That makes sense.
@civicrm-builder test this please.

**Review by colemanw — APPROVED** [review]

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
@colemanw does 'approved' == merge on pass?

**Comment by colemanw** [comment]
Approval + cat pic = merge on pass

**Comment by eileenmcnaughton** [comment]
@colemanw ah now I know :-)

**Pull request merged**